test/unit/math/prim/prob/hmm_marginal_test.cpp

```diff
@@ -0,0 +1,243 @@
+  for (int i = 0; i < n_states; i++) {
+    Gamma(i, n_states - 1) = 1 - sum(row(Gamma_unconstrained, i + 1));
+    for (int j = 0; j < n_states - 1; j++) {
+      Gamma(i, j) = Gamma_unconstrained(i, j);
+    }
+  }
+
+  Eigen::Matrix<T_rho, Eigen::Dynamic, 1> rho(n_states);
+  rho(1) = 1 - sum(rho_unconstrained);
+  for (int i = 0; i < n_states - 1; i++)
+    rho(i) = rho_unconstrained[i];
+
+  return stan::math::hmm_marginal_lpdf(log_omegas, Gamma, rho);
+}
+
+/**
+ * In the proposed example, the latent sate x determines
+ * the observational distribution:
+ *  (i) normal(mu, sigma)
+ *  (ii) normal(-mu, sigma)
+ */
+double state_lpdf(double y, double abs_mu, double sigma, int state) {
+  int x = (-2 * state + 1);
+  double chi = (y - x * abs_mu) / sigma;
+  return -0.5 * chi * chi - 0.5 * std::log(6.283185307179586) - std::log(sigma);
+}
+
+class hmm_marginal_lpdf_test : public ::testing::Test {
+ protected:
+  void SetUp() override {
+    n_states = 2;
+    p1_init = 0.65;
+    gamma1 = 0.7;
+    gamma2 = 0.45;
+    n_transitions = 10;
+    abs_mu = 1;
+    sigma = 1;
+
+    Eigen::VectorXd rho_(n_states);
+    rho_ << p1_init, 1 - p1_init;
+    rho = rho_;
+
+    Eigen::MatrixXd Gamma_(n_states, n_states);
+    Gamma_ << gamma1, 1 - gamma1, gamma2, 1 - gamma2;
+    Gamma = Gamma_;
+
+    Eigen::VectorXd obs_data_(n_transitions + 1);
+    obs_data_ << -0.3315914, -0.1655340, -0.7984021, 0.2364608, -0.4489722,
+        2.1831438, -1.4778675, 0.8717423, -1.0370874, 0.1370296, 1.9786208;
+    obs_data = obs_data_;
+
+    Eigen::MatrixXd log_omegas_(n_states, n_transitions + 1);
+    for (int n = 0; n < n_transitions + 1; n++) {
+      log_omegas_.col(n)[0] = state_lpdf(obs_data[n], abs_mu, sigma, 0);
+      log_omegas_.col(n)[1] = state_lpdf(obs_data[n], abs_mu, sigma, 1);
+    }
+    log_omegas = log_omegas_;
+    log_omegas_zero = log_omegas.block(0, 0, n_states, 1);
+
+    std::vector<double> rho_unconstrained_(n_states - 1);
+    for (int i = 0; i < rho.size() - 1; i++)
+      rho_unconstrained_[i] = rho(i);
+    rho_unconstrained = rho_unconstrained_;
+
+    Gamma_unconstrained = Gamma.block(0, 0, n_states, n_states - 1);
+  }
+
+  int n_states, n_transitions;
+  double abs_mu, sigma, p1_init, gamma1, gamma2;
+
+  Eigen::VectorXd rho;
+  Eigen::MatrixXd Gamma;
+  Eigen::VectorXd obs_data;
+  Eigen::MatrixXd log_omegas;
+  Eigen::MatrixXd log_omegas_zero;
+
+  // Construct "unconstrained" versions of rho and Gamma, without
+  // the final element which can be determnied using the fact
+  // the columns sum to 1. This allows us to do finite diff tests,
+  // without violating the simplex constraint of rho and Gamma.
+  std::vector<double> rho_unconstrained;
+  Eigen::MatrixXd Gamma_unconstrained;
+  stan::test::ad_tolerances tols;
+};
+
+TEST_F(hmm_marginal_lpdf_test, ten_transitions) {
+  using stan::math::hmm_marginal_lpdf;
+
+  EXPECT_FLOAT_EQ(-18.37417, hmm_marginal_lpdf(log_omegas, Gamma, rho));
```

Comment:
[question]
Where do these values come from?

Reply:
I wrote and ran the forward algorithm in R. I can share the script. In the same algorithm that we use, but the code is drastically simpler.

Reply:
Please just note that with a comment.  I'd suggest committing the R script, too, but I have no idea where it would go as it's not C++ or Stan.

Reply:
I'll bring it up at the next Stan meeting. For now, I'll leave the following comment:
```
// For evaluation of the density, the C++ code is benchmarked against
// a forward algorithm written in R.
// TODO(charlesm93): Add public repo link with R script. 
```

Reply:
OK.